### DISCUSSION TITLE
Is there a way to set site permissions?

### DISCUSSION THREAD

**Discussion by brian-kephart** [rubycdp/ferrum#377] [General]
Thanks for this great library!

My specific use case for altering site permissions is that I am testing a subscription workflow for push notifications, and I would like to test the user experience when the value of `Notification.permission` is `"default"`, `"granted"`, or `"denied"`.

```ruby
# in headless mode
browser.evaluate("Notification.permission") # => "denied"

# in non-headless mode
browser.evaluate("Notification.permission") # => "default"
```

I'm wondering if there is a way to set this value to `"granted"`, or to modify site permissions in general. I could not find anything relevant to this in the README, the code, or the command line args.

**Comment by route** [comment]
@brian-kephart hey, you could try set permissions with https://chromedevtools.github.io/devtools-protocol/tot/Browser/#method-grantPermissions

```
browser.command("Browser.grantPermissions", permissions: ['notifications'], origin: 'https://example.com')
```

If that works please leave feedback or maybe send a PR ;)

**Comment by brian-kephart** [comment]
Thanks for the tip! I still can't change the permissions, though.

Something is happening, because the commands I tried (`grantPermissions`, `setPermission`, and `resetPermissions`) raise exceptions unless the arguments conform to the spec at the link you showed. However, even when the commands are accepted, the permission available in the browser still does not change.

```ruby
browser.command "Browser.grantPermissions", permissions: ["notifications"] # returns {}
browser.evaluate "Notification.permission" # still "denied"
```

**Comment by route** [comment]
looks like a Chrome issue

https://github.com/puppeteer/puppeteer/issues/3279
https://github.com/puppeteer/puppeteer/issues/3241

**Comment by apmiller108** [comment]
I'm attempting to do something similar, but with `clipboardReadWrite`
```ruby
page.driver.browser.command('Browser.grantPermissions', origin: page.server_url, permissions: ['clipboardReadWrite'])
```

This, however throws the following error:

```
Failure/Error: page.driver.browser.command('Browser.grantPermissions', origin: page.server_url, permissions: ['clipboardReadWrite'])
     
Ferrum::BrowserError:
   Permissions can't be granted in current context.
```

**Reply by route** [reply]
Apparently it's not fixed upstream
